rct229/rules/section22/section22rule8.py

```diff
@@ -0,0 +1,141 @@
+from rct229.utils.pint_utils import ZERO
+
+APPLICABLE_SYS_TYPES = [
+    HVAC_SYS.SYS_7,
+    HVAC_SYS.SYS_8,
+    HVAC_SYS.SYS_11_1,
+    HVAC_SYS.SYS_11_2,
+    HVAC_SYS.SYS_12,
+    HVAC_SYS.SYS_13,
+    HVAC_SYS.SYS_7B,
+    HVAC_SYS.SYS_8B,
+    HVAC_SYS.SYS_11_1B,
+    HVAC_SYS.SYS_12B,
+]
+PUMP_SPPED_CONTROL = schema_enums["PumpSpeedControlOptions"]
+FLUID_LOOP = schema_enums["FluidLoopOptions"]
+MIN_CHW_PRIMARY_LOOP_COOLING_CAPACITY = 300.0 * ureg("ton")
+
+
+class Section22Rule8(RuleDefinitionListIndexedBase):
+    """Rule 8 of ASHRAE 90.1-2019 Appendix G Section 22 (Chilled water loop)"""
+
+    def __init__(self):
+        super(Section22Rule8, self).__init__(
+            rmrs_used=UserBaselineProposedVals(False, True, False),
+            each_rule=Section22Rule8.ChillerFluidLoopRule(),
+            index_rmr="baseline",
+            id="22-8",
+            description="For Baseline chilled water system with cooling capacity of 300 tons or more, the secondary pump shall be modeled with variable-speed drives.",
+            list_path="ruleset_model_instances[0].fluid_loops[*]",
+        )
+
+    def is_applicable(self, context, data=None):
+        rmr_baseline = context.baseline
+        rmi_b = rmr_baseline["ruleset_model_instances"][0]
+        baseline_system_types_dict = get_baseline_system_types(rmi_b)
+        # create a list contains all HVAC systems that are modeled in the rmi_b
+        available_type_lists = [
+            hvac_type
+            for hvac_type in baseline_system_types_dict.keys()
+            if len(baseline_system_types_dict[hvac_type]) > 0
+        ]
+
+        primary_secondary_loop_dictionary = get_primary_secondary_loops_dict(rmi_b)
+
+        return (
+            any(
+                [
+                    available_type in APPLICABLE_SYS_TYPES
+                    for available_type in available_type_lists
+                ]
+            )
+            and primary_secondary_loop_dictionary
+        )
+
+    def create_data(self, context, data):
+        rmr_baseline = context.baseline
+        rmi_b = rmr_baseline["ruleset_model_instances"][0]
+        loop_pump_dictionary = {
+            pump["loop_or_piping"]: pump for pump in find_all("pumps[*]", rmi_b)
+        }
+
+        chw_loop_capacity_dict = {}
+        for chiller in find_all("chillers[*]", rmi_b):
+            if chiller["cooling_loop"] not in chw_loop_capacity_dict.keys():
+                chw_loop_capacity_dict[chiller["cooling_loop"]] = ZERO.POWER
+            chw_loop_capacity_dict[chiller["cooling_loop"]] += getattr_(
+                chiller, "chiller", "rated_capacity"
+            )
+
+        primary_secondary_loop_dictionary = get_primary_secondary_loops_dict(rmi_b)
+
+        child_loop_speed_control_dict = {}
+        for loop_b_id in primary_secondary_loop_dictionary.keys():
+            for child_loop in find_one_with_field_value(
+                "fluid_loops[*]", "id", loop_b_id, rmi_b
+            )["child_loops"]:
```

Comment:
@charliepnnl I'm not sure if I understood correctly. Could you double-check this part again?

The way I implemented was i) to use ```find_all(f'$.fluid_loops[*][?(@.id="{primary_loop_id}")].child_loops[*]', rmi_b)``` ii) grab ```speed_control``` value using ```getattr_```.

Reply:
I think one step is missing in your implementation @yunjoonjung-PNNL .
The reason we use get_exactly_one_fluid_loop() function is trying to capture the possible errors if `loop_b_id` is not a real `fluid_loop` id. We want to capture the error and return to UNDETERMINED decision.

Using `find_all` on `child_loops` to find list of child_loops. However, it is OK for a fluid loop not having `child_loops`.

So in that case I think this can be rewritten as:
```python
for child_loop in find_all("$.child_loops[*]", find_exactly_one_with_field_value("$.fluid_loops[*]", "id", loop_b_id, rmi_b))

```